core/trino-main/src/main/java/io/trino/execution/scheduler/policy/SequentialExecutionSchedule.java

```diff
@@ -0,0 +1,606 @@
+        {
+            return lazyFragments;
+        }
+
+        public boolean isCurrentFragmentLazy()
+        {
+            return currentFragmentLazy;
+        }
+    }
+
+    private static StageExecution createStageExecution(PipelinedStageExecution stageExecution)
+    {
+        return new StageExecution()
+        {
+            @Override
+            public PlanFragment getFragment()
+            {
+                return stageExecution.getFragment();
+            }
+
+            @Override
+            public boolean isAnyTaskBlocked()
+            {
+                return stageExecution.isAnyTaskBlocked();
+            }
+
+            @Override
+            public State getState()
+            {
+                return stageExecution.getState();
+            }
+
+            @Override
+            public void addStateChangeListener(StateChangeListener<State> stateChangeListener)
+            {
+                stageExecution.addStateChangeListener(stateChangeListener);
+            }
+
+            @Override
+            public PipelinedStageExecution getPipelinedStageExecution()
+            {
+                return stageExecution;
+            }
+        };
+    }
+
+    private static class FragmentsEdgeFactory
+            implements EdgeFactory<PlanFragmentId, FragmentsEdge>
+    {
+        @Override
+        public FragmentsEdge createEdge(PlanFragmentId sourceVertex, PlanFragmentId targetVertex)
+        {
+            return new FragmentsEdge(sourceVertex, targetVertex);
+        }
+    }
+
+    @VisibleForTesting
+    static class FragmentsEdge
+    {
+        private final PlanFragmentId source;
+        private final PlanFragmentId target;
+
+        public FragmentsEdge(PlanFragmentId source, PlanFragmentId target)
+        {
+            this.source = requireNonNull(source, "source is null");
+            this.target = requireNonNull(target, "target is null");
+        }
+
+        public PlanFragmentId getSource()
+        {
+            return source;
+        }
+
+        public PlanFragmentId getTarget()
+        {
+            return target;
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("source", source)
+                    .add("target", target)
+                    .toString();
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            FragmentsEdge that = (FragmentsEdge) o;
+            return source.equals(that.source) && target.equals(that.target);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(source, target);
+        }
+    }
+
+    @VisibleForTesting
+    interface StageExecution
```

Comment:
Add a note that whole purpose of the interface is to be able to have a mock implementation in tests.
Did you consider making `PipelinedStageExecution` itself an interface?